test/integration/isolated/suite_test.go

```diff
@@ -0,0 +1,346 @@
+	cfg.WithKubeconfigFile(conf.ResolveKubeConfigFile())
+	tenv = env.NewWithConfig(cfg)
+
+	var (
+		// Specifying a run ID so that multiple runs wouldn't collide.
+		runID = RandomName("", 3)
+		// The env is shared and built only once.
+		env environments.Environment
+	)
+
+	fmt.Printf("INFO: runID %s\n", runID)
+
+	builder := environments.NewBuilder()
+	fmt.Println("INFO: configuring cluster for testing environment")
+	if existingCluster := testenv.ExistingClusterName(); existingCluster != "" {
+		if cv := testenv.ClusterVersion(); cv != "" {
+			helpers.ExitOnErrWithCode(ctx,
+				fmt.Errorf("can't flag cluster version (%s) & provide an existing cluster at the same time", cv),
+				consts.ExitCodeIncompatibleOptions)
+		}
+		clusterParts := strings.Split(existingCluster, ":")
+		if len(clusterParts) != 2 {
+			helpers.ExitOnErrWithCode(ctx, fmt.Errorf("existing cluster in wrong format (%s): format is <TYPE>:<NAME> (e.g. kind:test-cluster)", existingCluster), consts.ExitCodeCantUseExistingCluster)
+		}
+		clusterType, clusterName := clusterParts[0], clusterParts[1]
+
+		fmt.Printf("INFO: using existing %s cluster %s\n", clusterType, clusterName)
+		switch clusterType {
+		case string(kind.KindClusterType):
+			cluster, err := kind.NewFromExisting(clusterName)
+			helpers.ExitOnErr(ctx, err)
+			builder.WithExistingCluster(cluster)
+			builder.WithAddons(metallb.New())
+		default:
+			helpers.ExitOnErrWithCode(ctx, fmt.Errorf("unknown cluster type: %s", clusterType), consts.ExitCodeCantUseExistingCluster)
+		}
+
+	} else {
+		fmt.Println("INFO: no existing cluster found, deploying using Kubernetes In Docker (KIND)")
+
+		builder.WithAddons(metallb.New())
+
+		if testenv.ClusterVersion() != "" {
+			clusterVersion, err := semver.Parse(strings.TrimPrefix(testenv.ClusterVersion(), "v"))
+			helpers.ExitOnErr(ctx, err)
+
+			fmt.Printf("INFO: build a new KIND cluster with version %s\n", clusterVersion.String())
+			builder.WithKubernetesVersion(clusterVersion)
+		}
+	}
+
+	fmt.Println("INFO: building test environment")
+	env, err = builder.Build(ctx)
+	helpers.ExitOnErr(ctx, err)
+
+	fmt.Printf("INFO: waiting for cluster %s and all addons to become ready\n", env.Cluster().Name())
+	envReadyCtx, envReadyCancel := context.WithTimeout(ctx, testenv.EnvironmentReadyTimeout())
+	defer envReadyCancel()
+	helpers.ExitOnErr(ctx, <-env.WaitForReady(envReadyCtx))
+
+	if err := testutils.PrepareClusterForRunningControllerManager(ctx, env.Cluster()); err != nil {
+		helpers.ExitOnErr(ctx, fmt.Errorf("failed to prepare cluster for running the controller manager: %w", err))
+	}
+
+	ctx = SetClusterInCtx(ctx, env.Cluster())
+	ctx = SetRunIDInCtx(ctx, runID)
+	tenv = tenv.WithContext(ctx)
+
+	// TODO: can't use any of AfterEachFeature,BeforeEachFeature,AfterEachTest,BeforeEachTest
+	// to get conditional setup and teardown.
+	// Related: https://github.com/Kong/kubernetes-ingress-controller/issues/4847
```

Comment:
https://github.com/kubernetes-sigs/e2e-framework/issues/330 is already fixed so as I understand we could already use the `Before/AfterEach`  hooks, couldn't we?

Reply:
That's true. How about we address that in a separate PR?

`AfterEachFeature`,`BeforeEachFeature`,`AfterEachTest` and `BeforeEachTest` have different signatures than what's already in this PR here (use in `WithSetup()` https://github.com/Kong/kubernetes-ingress-controller/blob/dc77459d8f4e4a73f439518bc694f1b088070258/test/integration/isolated/udproute_test.go#L92). Additionally this would probably require us to update `TESTING.md` as well 😬 

We have https://github.com/Kong/kubernetes-ingress-controller/issues/4847 to track this.

WDYT?